README.md
# RLOG Client

Small client to receive, print and filter messages from remote devices running an [RLOG Server](https://github.com/eduardodsp/rlog)

## Requirements

- python (3.x.y)
- pip

## Installation
```bash
pip install -r requirements.txt
```

## How To Use
To use this client all you have to do is run the following command
```
python rlogcli.py -ip <rlog server ip address> -p <user defined tcp port>
```
The client will immediately start running and will try to connect to the server at the defined
ip address + port. If successful the client will begin to print all received messages in order.

### Example:
```
python rlogcli.py -ip 192.168.178.210 -p 8888
```
[![gif with examples][example-link]][example-link]

### Command Options
```
usage: rlogcli.py [-h] [-ip IP] [-p PORT] [-tag TAG] [-type TYPE] [-o OUTPUT]        

[RLOG] Remote Logger client. Receive, filter and save log messages from remote devices

options:
  -h, --help  show this help message and exit
  -ip IP      Server IP address
  -p PORT     TCP PORT used by target server
  -tag TAG    Tag filter. User defined message tag
  -type TYPE  Type filter. Available options: 1 = Info, 2 = Warnings, 3 = Errors      
  -o OUTPUT   Output filename
```
Arguments -tag and -type are display filters that allows the user to select which messages are to be printed on the terminal: 
- Tag: This is a user defined string that typically indicates the context of the message.
- Type: Indicates whether this is an Informational, Warning or Error message.

Argument -o allows the user to write all messages to a plain text file. By desgin all messages are saved to the file, even if display filters are being used. 

[example-link]:   https://github.com/eduardodsp/rlogcli/blob/main/example.gif

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0] - 2022-09-29

### Added
- CHANGELOG.md
- README.md
- requirements.txt
- rlogcli.py

[Unreleased]: https://github.com/eduardodsp/rlogcli/releases
[1.0.0]: https://github.com/eduardodsp/rlogcli/releases/tag/v1.0.0


rlog_cli.py
import sys
import argparse
import socket
import select
from pythonping import ping
from colorama import init, Fore, Back, Style

def main(argv):
   
    RLOG_BANNER = """ ______     __         ______     ______    
/\  == \   /\ \       /\  __ \   /\  ___\   
\ \  __<   \ \ \____  \ \ \/\ \  \ \ \__ \  
 \ \_\ \_\  \ \_____\  \ \_____\  \ \_____\ 
  \/_/ /_/   \/_____/   \/_____/   \/_____/ """

    # Initializes Colorama
    init(autoreset=True)

    parser = argparse.ArgumentParser(description='[RLOG] Remote Logger client. Receive, filter and save log messages from remote devices')

    parser.add_argument('-ip', type=str, default = "192.168.178.210", help='Server IP address')
    parser.add_argument('-p', dest = 'port', type=int, default = 8888, help='TCP PORT used by target server')
    parser.add_argument('-tag', type=str, help='Tag filter. User defined message tag')
    parser.add_argument('-type', type=int, help='Type filter. Available options: 1 = Info, 2 = Warnings, 3 = Errors')
    parser.add_argument('-o', dest = 'output', type=str, help='Output filename')

    args = parser.parse_args()
    
    HOST = args.ip
    PORT = args.port
    FILTER_TAG = args.tag
    FILTER_TYPE = args.type
    OUTPUT_FILE = args.output

    print(RLOG_BANNER)
    print("\n")
    
    MSG_HOST_PORT = HOST + ":" + str(PORT)
    MSG_WAITING_ON_SERVER = "Waiting for RLOG server on " + MSG_HOST_PORT
    MSG_CONNECTED_TO_SERVER = Fore.GREEN + "Connected to server on " + MSG_HOST_PORT + "!"
    MSG_DISCONNECTED_FROM_SERVER = Fore.RED + "Lost connection from server on " + MSG_HOST_PORT + "!"
    MSG_SERVER_UNREACHABLE = Fore.RED + "Server address is unreachable"
    MSG_FATAL_ERROR = Fore.RED + "Unexpected error!"

    exit = False
    attempts = 0

    print (MSG_WAITING_ON_SERVER)
    while (exit == False):
        try:
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:        
                try:
                    s.settimeout(10)
                    s.connect((HOST, PORT))
                    print (MSG_CONNECTED_TO_SERVER)
                    s.setblocking(0)
                    socket_list = [s]
                    connected = True
                    attempts = 0
                    while(connected) : 

                        read_sockets, write_sockets, error_sockets = select.select(socket_list, [], [], 5)

                        if len(read_sockets) != 0:
                            raw = s.recv(1024)
                            lmsg = str(raw.decode('ascii'))
                            if (len(lmsg) > 0):
                                process_msg(lmsg, FILTER_TAG, FILTER_TYPE)      
                                save_to_file(lmsg, OUTPUT_FILE)                      
                        else:
                            rsp = ping(HOST, verbose=False)
                            if (rsp.success() == False):
                                print (MSG_DISCONNECTED_FROM_SERVER)
                                connected = False

                except socket.timeout:
                    # timeout trying to connect, lets try some more times            
                    print (MSG_WAITING_ON_SERVER)    

                    if (attempts >= 2):
                        print (MSG_SERVER_UNREACHABLE)
                        exit = True

                    attempts = attempts + 1        

                except KeyboardInterrupt:
                    exit = True

        except:
            print (MSG_FATAL_ERROR)
            exit = True

    print ("Terminating RLOG client!")


def process_msg(msg, ftag, ftype):

    # apply tag filter
    if (ftag is not None):
        if (msg.find(str("["+ ftag + "]")) == -1):
            return None

    # get message type
    msg_type = get_msg_type(msg)

    # apply type filter
    if (ftype is not None):
        if (msg_type != ftype):
            return None

    # print filtered message
    print_colored(msg, msg_type)


def get_msg_type(msg):
    
    if (msg.find("[i]") > 0):
        return 1
    
    if (msg.find("[!]") > 0):
        return 2

    if (msg.find("[#]") > 0):
        return 3

    else:
        return None


def print_colored(msg, itype):

    if (itype == 1):
        print (msg, end="")
    
    elif (itype == 2):
        print (Fore.YELLOW + msg, end="")

    elif (itype == 3):
        print (Fore.RED + msg, end="")


    
def save_to_file(msg, outputfile):

    if(outputfile is not None):
        with open(outputfile, 'a') as f:
            f.write(msg)

if __name__ == '__main__':
    try:
        main(sys.argv[1::])
    except KeyboardInterrupt as ex:
        exit(1)



rlogcli.py
import sys
import argparse
import socket
import select
from pythonping import ping
from colorama import init, Fore, Back, Style

def main(argv):
   
    RLOG_BANNER = """ ______     __         ______     ______    
/\  == \   /\ \       /\  __ \   /\  ___\   
\ \  __<   \ \ \____  \ \ \/\ \  \ \ \__ \  
 \ \_\ \_\  \ \_____\  \ \_____\  \ \_____\ 
  \/_/ /_/   \/_____/   \/_____/   \/_____/ 
                                    v1.0.0"""

    # Initializes Colorama
    init(autoreset=True)

    parser = argparse.ArgumentParser(description='[RLOG] Remote Logger client. Receive, filter and save log messages from remote devices')

    parser.add_argument('-ip', type=str, default = "192.168.178.210", help='Server IP address')
    parser.add_argument('-p', dest = 'port', type=int, default = 8888, help='TCP PORT used by target server')
    parser.add_argument('-tag', type=str, help='Tag filter. User defined message tag')
    parser.add_argument('-type', type=int, help='Type filter. Available options: 1 = Info, 2 = Warnings, 3 = Errors')
    parser.add_argument('-o', dest = 'output', type=str, help='Output filename')

    args = parser.parse_args()
    
    HOST = args.ip
    PORT = args.port
    FILTER_TAG = args.tag
    FILTER_TYPE = args.type
    OUTPUT_FILE = args.output

    print(RLOG_BANNER)
    print("\n")
    
    MSG_HOST_PORT = HOST + ":" + str(PORT)
    MSG_WAITING_ON_SERVER = "Waiting for RLOG server on " + MSG_HOST_PORT
    MSG_CONNECTED_TO_SERVER = Fore.GREEN + "Connected to server on " + MSG_HOST_PORT + "!"
    MSG_DISCONNECTED_FROM_SERVER = Fore.RED + "Lost connection from server on " + MSG_HOST_PORT + "!"
    MSG_SERVER_UNREACHABLE = Fore.RED + "Server address is unreachable"
    MSG_FATAL_ERROR = Fore.RED + "Unexpected error!"

    exit = False
    attempts = 0

    print (MSG_WAITING_ON_SERVER)
    while (exit == False):
        try:
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:        
                try:
                    s.settimeout(10)
                    s.connect((HOST, PORT))
                    print (MSG_CONNECTED_TO_SERVER)
                    s.setblocking(0)
                    socket_list = [s]
                    connected = True
                    attempts = 0
                    while(connected) : 

                        read_sockets, write_sockets, error_sockets = select.select(socket_list, [], [], 5)

                        if len(read_sockets) != 0:
                            raw = s.recv(1024)
                            lmsg = str(raw.decode('ascii'))
                            if (len(lmsg) > 0):
                                process_msg(lmsg, FILTER_TAG, FILTER_TYPE)      
                                save_to_file(lmsg, OUTPUT_FILE)                      
                        else:
                            rsp = ping(HOST, verbose=False)
                            if (rsp.success() == False):
                                print (MSG_DISCONNECTED_FROM_SERVER)
                                connected = False

                except socket.timeout:
                    # timeout trying to connect, lets try some more times            
                    print (MSG_WAITING_ON_SERVER)    

                    if (attempts >= 2):
                        print (MSG_SERVER_UNREACHABLE)
                        exit = True

                    attempts = attempts + 1        

                except KeyboardInterrupt:
                    exit = True

        except:
            print (MSG_FATAL_ERROR)
            exit = True

    print ("Terminating RLOG client!")


def process_msg(msg, ftag, ftype):

    # apply tag filter
    if (ftag is not None):
        if (msg.find(str("["+ ftag + "]")) == -1):
            return None

    # get message type
    msg_type = get_msg_type(msg)

    # apply type filter
    if (ftype is not None):
        if (msg_type != ftype):
            return None

    # print filtered message
    print_colored(msg, msg_type)


def get_msg_type(msg):
    
    if (msg.find("[i]") > 0):
        return 1
    
    if (msg.find("[!]") > 0):
        return 2

    if (msg.find("[#]") > 0):
        return 3

    else:
        return None


def print_colored(msg, itype):

    if (itype == 1):
        print (msg, end="")
    
    elif (itype == 2):
        print (Fore.YELLOW + msg, end="")

    elif (itype == 3):
        print (Fore.RED + msg, end="")


    
def save_to_file(msg, outputfile):

    if(outputfile is not None):
        with open(outputfile, 'a') as f:
            f.write(msg)

if __name__ == '__main__':
    try:
        main(sys.argv[1::])
    except KeyboardInterrupt as ex:
        exit(1)

